crates/bevy_ecs/macros/src/world_query.rs

```diff
@@ -0,0 +1,179 @@
+use proc_macro2::Ident;
+use quote::quote;
+use syn::{Attribute, Fields, ImplGenerics, TypeGenerics, Visibility, WhereClause};
+
+#[allow(clippy::too_many_arguments)]
+pub(crate) fn item_struct(
+    path: &syn::Path,
+    fields: &Fields,
+    derive_macro_call: &proc_macro2::TokenStream,
+    struct_name: &Ident,
+    visibility: &Visibility,
+    item_struct_name: &Ident,
+    field_types: &Vec<proc_macro2::TokenStream>,
+    user_impl_generics_with_world: &ImplGenerics,
+    field_attrs: &Vec<Vec<Attribute>>,
+    field_visibilities: &Vec<Visibility>,
+    field_idents: &Vec<proc_macro2::TokenStream>,
+    user_ty_generics: &TypeGenerics,
+    user_ty_generics_with_world: &TypeGenerics,
+    user_where_clauses_with_world: Option<&WhereClause>,
+) -> proc_macro2::TokenStream {
```

Comment:
Is there any way to avoid the sheer number of arguments here and below?

Reply:
Quite possibly. I will have a go at factoring all the arguments into a small number of structs instead.